reactor-test/src/main/java/reactor/test/publisher/ColdTestPublisher.java

```diff
@@ -226,6 +221,76 @@ void onComplete() {
 			parent.remove(this);
 			actual.onComplete();
 		}
+
+		private void drain(long n) {
+			int i = index;
+			int emitted = 0;
+			for ( ; ; ) {
+				if (cancelled) {
+					return;
+				}
+
+				while (i != parent.values.size()) {
+					if (emitted == n && !parent.violations.contains(REQUEST_OVERFLOW)) {
+						break;
+					}
+					T t = parent.values.get(i);
+					if (t == null && !parent.violations.contains(ALLOW_NULL)) {
+						actual.onError(new NullPointerException("The " + i + "th element was null"));
+						return;
+					}
+
+					if (this.onNext(t)) {
+						emitted++;
+					}
+					i++;
+					if (cancelled) {
+						return;
+					}
+				}
+
+				n = requested;
+				if (emitted > n) {
+					assert parent.violations.contains(Violation.REQUEST_OVERFLOW);
+					parent.hasOverflown = true;
+					return;
```

Comment:
shouldn't it check for `emitTerminalSignalIfAny()` here too?

Reply:
It could indeed. That branch is the very exceptional case I thought, leaving the publisher in a "bad state" state, but it could go the extra mile and also emit terminal signals (which should be also protected against)